docs/debug_tests.rst

```diff
@@ -0,0 +1,81 @@
+        for hit in res['hits']['hits']:
+            assert 'kimchy’ == hit['_source']['author’], "Expected author kimchy but got %s" % hit['_source']['author']
+            assert 'Elasticsearch: cool.' == hit['_source']['text’], "Expected text Elasticsearch: cool. but got %s" % hit['_source']['text’]
+
+First the tests outputs the result of a search, so that if any of the following assertions fail, we can see the whole result in ``test_log.debug``. Assertion messages help to quickly see the difference in expected and retrieved results. 
+
+
+Fail early
+==========
+
+Try to avoid a situation where a test fails because of an uncaught failure earlier in the test. Suppose we write a ``start_node`` method that does not check if the service starts successfully. The service fails to start, but we get a test failure indication that there was a problem querying the service. It would be much faster to debug the issue if the test failure pointed to the issue with starting the service. So make sure to add checks for operations that may fail, and fail the test earlier than later.
+
+
+Flaky tests
+============
+
+Flaky tests are hard to debug due to their non-determinism, they waste time, and sometimes hide real bugs: developers tend to ignore those failures, and thus could miss real bugs. Flakiness can come from the test itself, the system it is testing, or the environmental issues.
+
+Asynchronous Wait
+^^^^^^^^^^^^^^^^^
+
+A common cause of a flaky test is asynchronous wait. A test makes an asynchronous call and does not properly wait for the result of the call to become available before using it::
+
+	node.account.kill_process("zookeeper", allow_fail=False)
+	time.sleep(2)
+	assert not self.alive(node), “Expected Zookeeper service to stop” 
+
+In this example, the test terminates a zookeeper service via ``kill_process`` and then uses ``time.sleep`` to wait for it to stop. If terminating the process takes longer, the test will fail. The test may intermittently fail based on how fast a process terminates. Of course, there should be a timeout for termination to ensure that test does not run indefinitely. You could increase sleep time, but that also increases the test run length. A more explicit way to express this condition is to use :meth:`~ ducktape.utils.util.wait_until` with a timeout::
+
+	node.account.kill_process("zookeeper", allow_fail=False)
+	wait_until(lambda: not self.alive(node), timeout_sec=5, err_msg="Timed out waiting for zookeeper to stop.")
+
+The test will progress as soon as condition is met, and timeout ensures that the test does not run indefinitely if termination never ends.
+
+Test Order Dependency
+^^^^^^^^^^^^^^^^^^^^^
+
+Make sure that your services properly cleanup the state in ``clean_node`` implementation. This can cause the next run of the test to fail or fail intermittently if other tests happen to clean same directories for example.
```

Comment:
I'd also recommend that they clean up aggressively. One of the benefits of the isolation that ducktape assumes is that you can assume you have complete control of the machine, so killing all java processes you can find rather than being more targeted is safe, deleting the entire temporary working space is ok, etc.